File: DecompilerData/scripts/export_decomp.py
# Export all functions + decompiled C from Ghidra headless.
#@category THExport
#@runtime PyGhidra
import os, json

args = getScriptArgs()
outdir = args[0] if args else os.getcwd()
os.makedirs(outdir, exist_ok=True)

program = currentProgram
fm = program.getFunctionManager()
listing = program.getListing()

from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor

di = DecompInterface()
di.openProgram(program)
monitor = ConsoleTaskMonitor()

funcs = [f for f in fm.getFunctions(True)]
summary = []
c_out = open(os.path.join(outdir, 'decomp_all.c'), 'w', encoding='utf-8', errors='replace')
c_out.write('// Ghidra decompilation export: %s\n// total functions: %d\n\n' % (program.getName(), len(funcs)))

count = 0
for f in funcs:
    body = f.getBody()
    size = body.getNumAddresses()
    entry = f.getEntryPoint()
    callees = set()
    called = f.getCalledFunctions(monitor)
    for cf in called:
        callees.append if False else callees.add(cf.getName())
    summary.append({
        'name': f.getName(),
        'entry': str(entry),
        'size': int(size),
        'params': f.getParameterCount(),
        'callees': sorted(callees),
    })
    try:
        res = di.decompileFunction(f, 60, monitor)
        if res.decompileCompleted():
            code = res.getDecompiledFunction().getC()
        else:
            code = '// DECOMPILE FAILED: %s\n' % f.getName()
    except Exception as e:
        code = '// EXCEPTION: %s / %s\n' % (f.getName(), e)
    c_out.write('// ===== FUNC %s @ %s (size=%d) =====\n' % (f.getName(), entry, size))
    c_out.write(code + '\n\n')
    count += 1
    if count % 100 == 0:
        print('decompiled %d/%d' % (count, len(funcs)))

c_out.close()

with open(os.path.join(outdir, 'functions.json'), 'w', encoding='utf-8') as fp:
    json.dump(summary, fp, ensure_ascii=False, indent=1)

print('DONE. functions: %d -> %s' % (count, outdir))

File: DecompilerData/scripts/ExportDecomp.java
// Export all functions + decompiled C from Ghidra headless.
//@category THExport
import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.*;
import ghidra.program.model.listing.*;
import ghidra.program.model.symbol.*;
import ghidra.util.task.ConsoleTaskMonitor;
import java.io.*;
import java.util.*;

public class ExportDecomp extends GhidraScript {
    @Override
    public void run() throws Exception {
        String[] args = getScriptArgs();
        String outdir = args.length > 0 ? args[0] : getProjectRootFolder() != null ? "." : ".";
        new File(outdir).mkdirs();

        FunctionIterator it = currentProgram.getFunctionManager().getFunctions(true);
        List<Function> funcs = new ArrayList<>();
        while (it.hasNext()) funcs.add(it.next());
        println("total functions: " + funcs.size());

        DecompInterface di = new DecompInterface();
        DecompileOptions opts = new DecompileOptions();
        di.setOptions(opts);
        di.openProgram(currentProgram);
        ConsoleTaskMonitor mon = new ConsoleTaskMonitor();

        PrintWriter csv = new PrintWriter(new OutputStreamWriter(
            new FileOutputStream(new File(outdir, "functions.csv")), "UTF-8"));
        csv.println("entry,name,size,params,callees");

        PrintWriter cOut = new PrintWriter(new OutputStreamWriter(
            new FileOutputStream(new File(outdir, "decomp_all.c")), "UTF-8"));
        cOut.println("// Ghidra decompilation export: " + currentProgram.getName());
        cOut.println("// total functions: " + funcs.size());
        cOut.println();

        int count = 0;
        for (Function f : funcs) {
            long size = f.getBody().getNumAddresses();
            String entry = f.getEntryPoint().toString();
            List<String> callees = new ArrayList<>();
            Set<Function> called = f.getCalledFunctions(mon);
            for (Function cf : called) callees.add(cf.getName());
            Collections.sort(callees);
            csv.println(entry + "," + f.getName() + "," + size + "," + f.getParameterCount() + ",\"" + String.join(" ", callees) + "\"");

            String code;
            try {
                DecompileResults res = di.decompileFunction(f, 60, mon);
                if (res != null && res.decompileCompleted() && res.getDecompiledFunction() != null) {
                    code = res.getDecompiledFunction().getC();
                } else {
                    code = "// DECOMPILE FAILED: " + f.getName();
                }
            } catch (Exception e) {
                code = "// EXCEPTION: " + f.getName() + " / " + e;
            }
            cOut.println("// ===== FUNC " + f.getName() + " @ " + entry + " (size=" + size + ") =====");
            cOut.println(code);
            cOut.println();
            count++;
            if (count % 200 == 0) println("decompiled " + count + "/" + funcs.size());
        }
        csv.close();
        cOut.close();
        println("DONE. functions: " + count + " -> " + outdir);
    }
}
